dockstore-webservice/src/main/java/io/dockstore/webservice/languages/IpynbHandler.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2023 OICR, UCSC
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.dockstore.webservice.languages;
+
+import io.dockstore.common.DescriptorLanguage;
+import io.dockstore.common.VersionTypeValidation;
+import io.dockstore.webservice.core.Author;
+import io.dockstore.webservice.core.SourceFile;
+import io.dockstore.webservice.core.Version;
+import io.dockstore.webservice.core.Workflow;
+import io.dockstore.webservice.helpers.SourceCodeRepoInterface;
+import io.dockstore.webservice.jdbi.ToolDAO;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class provides support for Jupyter .ipynb notebooks.
+ * https://nbformat.readthedocs.io/en/latest/format_description.html
+ * https://repo2docker.readthedocs.io/en/latest/specification.html
+ */
+public class IpynbHandler implements LanguageHandlerInterface {
+    public static final Logger LOG = LoggerFactory.getLogger(IpynbHandler.class);
+
+    public static final Set<String> REES_FILES = Set.of("environment.yml", "Pipfile", "Pipfile.lock", "requirements.txt", "setup.py", "Project.toml", "REQUIRE", "install.R", "apt.txt", "DESCRIPTION", "postBuild", "start", "runtime.txt", "default.nix", "Dockerfile");
+    public static final Set<String> REES_DIRS = Set.of("/", "/binder/", "/.binder/");
+
+    @Override
+    public Version parseWorkflowContent(String filePath, String content, Set<SourceFile> sourceFiles, Version version) {
+
+        // Parse the notebook JSON.
+        JSONObject notebook;
```

Comment:
Is there any library for this to avoid working with raw json objects?
Is something like a swagger/openapi definition we can work with?

Reply:
> Is there any library for this to avoid working with raw json objects? Is something like a swagger/openapi definition we can work with?

Looked for something like this before writing the code, but didn't find anything.  The notebook format is pretty simple, the object hierarchy is shallow and there's no recursive features or anything like that.  So, maybe nobody ever experienced enough pain to implement it...

Reply:
Maybe you can grab https://github.com/apache/zeppelin/pull/2393/commits/655cc4d7eb0a4b6790da90fffdf861d66e35ec88 from https://github.com/apache/zeppelin/pull/2393

Reply:
> Maybe you can grab [apache/zeppelin@655cc4d](https://github.com/apache/zeppelin/commit/655cc4d7eb0a4b6790da90fffdf861d66e35ec88) (or the most recent version of it) from [apache/zeppelin#2393](https://github.com/apache/zeppelin/pull/2393)

So, I spent a decent amount of time integrating the latest apache/zeppelin-jupyter, which contains that utility class which parses the notebook JSON to a class hierarchy.  Everything was going well, things were building and running, until I got to the author parsing code, and the compiler complained about no `getName` method.  What?  Oh, the `Author` class doesn't have any getters, lol: https://github.com/apache/zeppelin/blob/master/zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/nbformat/Author.java

Thus, I scuttled the zeppelin-jupyter option and instead massaged the existing JSON parsing code to be a little more legible.

This could be a good thing, I get the sense there's more entropy in the Jupyter ecosystem than the CWL/WDL communities.  There's a tension between validating by completely parsing, which might accidentally reject some valid variants/extensions and/or newer versions, and doing nothing.  Possibly, _'it parses as valid JSON and has correctly-typed values for the "metadata", "nbformat", "nbformat_minor", and "cells" fields_' is a good balance of the extremes.

I addressed the other feedback, and will write the `SourceCodeRepoInterface` refactor ticket after this is merged.

Reply:
> What? Oh, the `Author` class doesn't have any getters

You can auto-generate those https://www.jetbrains.com/help/idea/generating-code.html#generate-getters-setters

> There's a tension between validating by completely parsing, which might accidentally reject some valid variants/extensions and/or newer versions

Maybe that's why Zepplin used GSON, ignores unknown fields https://www.baeldung.com/gson-deserialization-guide#deserialize-json-with-extra-unknown-fields-to-object so if there were newer versions with newer fields, it should be just fine. 

Something with a conflicting but completely different definition, I think it should be treated as a different language anyway.